File: 0011-container-with-most-water/0011-container-with-most-water.go
func maxArea(height []int) int {
    maxArea := 0
    left := 0
    right := len(height) - 1

    for left < right {
        // 1. Calculate current area
        w := right - left
        h := 0
        if height[left] < height[right] {
            h = height[left]
        } else {
            h = height[right]
        }
        
        currentArea := w * h
        
        // 2. Update global max
        maxArea = max(maxArea, currentArea)

        // 3. Move the pointer pointing to the shorter line
        // This is the greedy step to find a taller bottleneck
        if height[left] < height[right] {
            left++
        } else {
            right--
        }
    }

    return maxArea
}

File: 0011-container-with-most-water/0011-container-with-most-water.rs
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut max :i32 = 0;
        let lh = height.len();
        if lh < 2{
            return 0;
        }

        let mut left :usize = 0;
        let mut right :usize = lh-1;
        while left < right{
            let mut area :i32 = 0;
            if height[left] > height[right]{
                area = height[right] * (right-left) as i32;
                right -= 1;
            } else{
                area = height[left] * (right-left) as i32;
                left += 1;
            }

            if area > max{
                max =area;
            }
        }
        max
    }
}
